### PR TITLE
GitHub Action for Deployments

### PR DESCRIPTION
**Pull request by aepfli**
Adding GitHub Actions for Deployment which will run
on Linux and with JDK 8. On the side i removed a lot
of unnecessary gradle code for publishing.

To get this running within the base repository, we
need to add following secrets:
- GH_WRITE_TOKEN: for updating the repo and setting the tag
- BINTRAY_USER_NAME: for upload to bintray
- BINTRAY_API_KEY: for upload to bintray
- NEXUS_TOKEN_USERNAME: for syncing bintray to maven central
- NEXUS_TOKEN_PASSWORD: for syncing bintray to maven central

:warning: We need to be careful with this, as this can break the releasing process :warning: 

---

I hereby agree to the terms of the [JUnit Pioneer Contributor License Agreement](https://github.com/junit-pioneer/junit-pioneer/blob/master/CONTRIBUTING.md#junit-pioneer-contributor-license-agreement).

### CODE REVIEW

**Comment by Bukama** [comment]
If I apply my basic Jenkins pipeline knowledge this PR looks good to me.

**Comment by Bukama** [comment]
> If I apply my basic Jenkins pipeline knowledge this PR looks good to me.

I have to revert my word. Please update the documentation.

>> Like Mockito, JUnit Pioneer implements a continuous delivery model using Shipkit and Travis CI. 

This is no longer true. But I could live with it, when we decide / progress about #193

**Comment by aepfli** [comment]
> > If I apply my basic Jenkins pipeline knowledge this PR looks good to me.
> 
> I have to revert my word. Please update the documentation.
> 
> > > Like Mockito, JUnit Pioneer implements a continuous delivery model using Shipkit and Travis CI.
> 
> This is no longer true. But I could live with it, when we decide / progress about #193

actually i would wait till the whole travis migration is done, and put that into #192 as a cleanup step. currently we are in a slight transition phase, and we need to do some cleanup afterwards anyways. when the page sync is done

**Comment by nicolaiparlog** [comment]
I got the impression that this is blocked by #193, so I marked it as such.

**Comment by aepfli** [comment]
> I got the impression that this is blocked by #193, so I marked it as such.

the question is, do we favor getting away from travis more than evaluating shipkit. i owuld first proceed with migration of github actions, and than do the evaluation. anyway we need to adapt the buildscripts on a later stage depending on the outcome of shipkit evaluation. i would try to use one CI tool first currently.

**Comment by nicolaiparlog** [comment]
Yeah, I misunderstood the situation. We can go forward with this and change the ShipKit situation separately.

**Comment by nicolaiparlog** [comment]
Finally took some time to go through this. Added all the five tokens you mentioned above under the names you need them. But they all come from an old dusty safe and I have no idea whether they're exactly what you need (e.g. I have a Sonatype token in my safe - let's assume it's the NEXUS_TOKEN we need here) and even still work. Let's find out!

**Comment by aepfli** [comment]
@nicolaiparlog now comes the part, where i am not 100% sure how to handle that in the best way. How do i test it now, if everything works as intended on the pioneer repo. As i am quiet new to releasing with shipkit bintray and jcenter, i suggest we coordinate on this.

i do not want to start with creating new releases of junit-pioneer. Therefor if it is okay, i suggest i shortly will change the name of the artifact to `junit-pioneer-test` and deactivate the automatic sync.

We then can check the jars and poms, if everything is still in place, and i would suggest to continue after that, by setting the right version. we could theoretically create a second branch like `release/testBintray` which would use this branch as a basis. So the git history within this branch would not be filled with unnecessary changes.

So to summarize:
1. create a branch based on that with a naming, which would invoke a shipkit release
    - jcenter sync has to be deactivated in the initial commit
    - the artifact name should be overwritten, or the version should have something like `SNAPSHOT` in its name
2. trigger the pipeline which will result in a deployment
3. checking the files on bintray if they are in a good shape
    - correct pom entries
    - working jar
4. merge this pullrequest and delete the interims branch.

what do you think?

**Comment by aepfli** [comment]
okay so lets write a good commit message:

```
GitHub Action for Deployments 

Adding GitHub Actions for Deployment which will run
on Linux and with JDK 8. This step allows us to remove
travis completely from the project. Additionally we also
remove appveyor as it is also redundant with GitHub 
Actions now.

We additionally removed all references to travis and
replaced them with new refernces to GitHub Actions.

As the Sitebuild was also broken, we also removed this
implementation and we will in the next step start 
implementing a new SiteBuild with GitHub Actions, and
a trigger will than be added to the actions within this repo

Closes: #199 
PR: #213 
Relates: #192 

```

**Comment by Bukama** [comment]
> Closes: #199 
> PR: #213
> Relates: #192 

Also closes #234 , doesn't it ?

**Comment by aepfli** [comment]
> > Closes: #199
> > PR: #213
> > Relates: #192
> 
> Also closes #234 , doesn't it ?
> 
> Why is #192 not closed? For the site deployment there is an own issue in the repo?

which suitebuild issue, there is the one within the junitpioneer page, but in here, we also will need to configure the trigger, hence that i would not close #192 till this last piece is done

File: .github/workflows/actions.yml [hunk 1]
```diff
@@ -45,42 +47,44 @@ jobs:
...
+      - name: Setup java
+        uses: actions/setup-java@v1
+        with:
+          java-version: 8
+      - name: Perform Release
+        uses: eskatos/gradle-command-action@v1
+        env:
+          TRAVIS_BRANCH: ${{ steps.gitbranch.outputs.replaced }}
+          TRAVIS_COMMIT_MESSAGE: ${{ github.event.head_commit.message }}
+          TRAVIS_PULL_REQUEST: ${{ github.event_name == 'pull_request' }}
```
Comment (beatngu13) [inline .github/workflows/actions.yml:79]: What do we need these environment variables for?

File: build.gradle.kts [hunk 2]
```diff
@@ -148,67 +137,4 @@ tasks {
...
-    }
-}
-
-tasks.register<org.junitpioneer.gradle.TriggerTravisTask>("triggerSiteBuild") {
-    travisProject = "junit-pioneer/junit-pioneer.github.io"
-    branch = "grandmaster"
-    apiToken = travisApiToken
-    message = "Triggered by successful JUnit Pioneer build for %COMMIT"
-}
+}
```
Comment (beatngu13) [inline build.gradle.kts:140]:
```suggestion
}

```

File: .github/workflows/actions.yml [hunk 3]
```diff
@@ -45,42 +47,44 @@ jobs:
...
+          GH_WRITE_TOKEN: ${{ secrets.GH_WRITE_TOKEN }}
```
Comment (beatngu13) [inline .github/workflows/actions.yml:80]: What is the difference to `GITHUB_TOKEN` above? If there is none, I would make the environment variables globally available. Otherwise, I would use the same prefix for consistency reasons (i.e. "GH" or "GITHUB").

File: .github/workflows/actions.yml [hunk 4]
```diff
@@ -45,42 +47,44 @@ jobs:
...
-  #       with:
-  #         arguments: ciPerformRelease
-  #     - name: Upload Release to GitHub
-  #       uses: softprops/action-gh-release@v1
-  #       if: startsWith(github.ref, 'refs/tags/')
-  #       with:
-  #         files: ./build/libs/junit-pioneer-*.jar
-  #       env:
-  #         GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+  deploy:
```
Comment (beatngu13) [inline .github/workflows/actions.yml:56]: If we want to, we can also use a separate YAML file for this?

File: .github/workflows/actions.yml (hunk 1 above)
Comment (aepfli) [reply .github/workflows/actions.yml:79]: that is the funny part, shipkit is only compatible with travis currently, and needs those travis environment variables. So i "fake" to be travis to get shipkit working, till they hopefully accept my proposal, and i can finish https://github.com/mockito/shipkit/pull/868

File: .github/workflows/actions.yml (hunk 4 above)
Comment (aepfli) [reply .github/workflows/actions.yml:56]: i was also thinking about this, and maybe trigger this from the build script. but i think we should always first run the build, and see if there are no errors, and than deploy.

This might be a good improvement for the future to find a sophisticated way of running both actions in sequence but conditionally on a success of the first one. I did not investigate that far into actions.

File: .github/workflows/actions.yml (hunk 3 above)
Comment (aepfli) [reply .github/workflows/actions.yml:80]: good point @nicolaiparlog can you rename the secret and add edit this line afterwards?

The difference is, that the `GITHUB_TOKEN` should not have any write access just read access, where ass the (soon to be) `GITHUB_WRITE_TOKEN` also has write access

Comment (beatngu13) [reply .github/workflows/actions.yml]: Ah, I see. Then maybe `GITHUB_READ_TOKEN` and `GITHUB_WRITE_TOKEN`? (Although I wonder what a read-only token on a public project can do that I cannot do without a token at all?)

Comment (aepfli) [reply .github/workflows/actions.yml]: i renamed the environment variable, we just need to change the name of the secret

Comment (aepfli) [reply .github/workflows/actions.yml]: i would not rename the other one, as it seems to be a default not 100% assinged to the repo but also to each user https://help.github.com/en/actions/configuring-and-managing-workflows/authenticating-with-the-github_token

**Comment by aepfli** [comment]
this pr just recreates the status quo regarding deployment, there is no alternative  merge behaviour with tags included. This is something we do with #193 and we will handle that afterwards.

**Comment by nicolaiparlog** [comment]
Err, I can't add a `GITHUB_WRITE_TOKEN`. When I enter the needed information and hit _add Secret_, the _Add new secret_ form disappears, but there's no new token. I can add that same token under another name, but whatever token value I use, I simply can't create a token with that name. Weird...

**Comment by nicolaiparlog** [comment]
As I understand, the new token name is a cosmetic improvement and everything already works, right? In that case, I'll merge.

**Comment by aepfli** [comment]
just a sec, with merge please, i check if it works, if you do not rename

**Comment by aepfli** [comment]
good to go!
